test/xbgpu/test_engine.py

```diff
@@ -866,32 +917,69 @@ async def _send_data(
                     f"actual: {ig_recv['frequency'].value}."
                 )
 
-                corrprod_results[i][j] = ig_recv["xeng_raw"].value
+                corrprod_results[corrprod_output.name][j] = ig_recv["xeng_raw"].value
+
+        beam_results: dict[str, np.ndarray] = {}
+        if beam_capture_stop_heap_indices is not None:
+            # We adjust the corresponding beam stream to only receive the required
+            # amount of data (up until the capture-stop point).
+            for beam_output, capture_stop_heap_index in zip(beam_outputs, beam_capture_stop_heap_indices):
+                if capture_stop_heap_index > 0:
```

Comment:
It sounds like you're treating non-positive values as magic value meaning "let it run to the end". It would be more explicit None for this purpose, and declare the arguments as `list[int | None]`.